### PR TITLE
Convert Token from enum to struct

### LINKED ISSUE #255: Convert Token from enum to struct with type property
Currently, `Token` is represented as enum which does not allow to store any properties in it, so we `components` property is computed, which affects performance as they can be parsed several times. We should instead convert it to `struct` and use a property of enum type like `type`.

### PR DESCRIPTION
**Pull request by ilyapuchka**
Resolves #255. Strangely enough, it makes parsing slightly slower, but I still think it's a good thing to do.
I didn't want to change too much code so I left static constructors instead of cases for interface to be the same.

### CODE REVIEW

File: Sources/Tokenizer.swift [hunk 1]
```diff
@@ -71,47 +71,53 @@ public struct SourceMap: Equatable {
...
+    /// A token representing a template block.
+    case block
+  }
+  
+  public let contents: String
+  public let kind: Kind
+  public let sourceMap: SourceMap
+  
+  /// Returns the underlying value as an array seperated by spaces
+  public lazy var components: [String] = self.contents.smartSplit()
```
Comment (AliSoftware) [inline Sources/Tokenizer.swift]: Shouldn't that be `private(set)`?

Comment (djbe) [reply Sources/Tokenizer.swift]: jep

Comment (djbe) [reply Sources/Tokenizer.swift]: We may also want to avoid splitting into components if the type is text?

File: Sources/Tokenizer.swift [hunk 2]
```diff
@@ -71,47 +71,53 @@ public struct SourceMap: Equatable {
...
-  public var contents: String {
-    switch self {
-    case .block(let value, _),
-         .variable(let value, _),
-         .text(let value, _),
-         .comment(let value, _):
-      return value
-    }
+  
+  public static func == (lhs: Token, rhs: Token) -> Bool {
```
Comment (djbe) [inline Sources/Tokenizer.swift]: Don't we get equatable for free?

File: Sources/Tokenizer.swift (hunk 1 above)
Comment (ilyapuchka) [reply Sources/Tokenizer.swift]: We can return just one item with the whole content, but I think this should be handled on a higher level and might as well not happening at all due to the logic flow

File: Sources/Tokenizer.swift (hunk 2 above)
Comment (ilyapuchka) [reply Sources/Tokenizer.swift]: not for classes...

**Comment by djbe** [comment]
Quick question: how did you measure a slowdown?

**Comment by ilyapuchka** [comment]
in the performance test, it went from 250-something to 260-something

**Comment by djbe** [comment]
The performance test is always a bit inconsistent though, you have to run it a few times to get an idea if it's a fluke, or an actual increase.

Anyway, 10ms increase isn't a problem in this case.

**Comment by djbe** [comment]
What we're not noticing now, is a potential performance benefit in other parts of Stencil, as we only measure the lexer performance.

**Comment by djbe** [comment]
Code looks GTM.

Might be worth adding documentation like I did in https://github.com/stencilproject/Stencil/pull/226/commits/cb4e51484660d17c92c0baa8e4c6f1e1b0f8e1a0, the current codebase can be quite complicated for developers to understand, and the docs should help with that.

Also, don't forget a changelog entry! 😉

**Pull request merged**